### PR TITLE
Domain functionality.

### PR DESCRIPTION
**Pull request by dariavh**
**Short description**

This PR will provide functionalities to construct `Domain`-type objects to merge duplicate code in `GQCD` and `GQCC`. It is a first suggestion towards a new feature in `GQCP` to calculate `Domain`-related features such as Mulliken charges, MPDs, ...

**To do**

- [x] (if applicable, when creating new source files): don't forget to update the Doxygen file, the collective `gqcp.hpp` include header and the CMake files
- [x] Implement base class `SimpleDomain` with an element type `ElementType` that is characteristic to the inheritinc classes. 
- [x] Implement class `DiscreteDomain` as a child class of `SimpleDomain`. It serves as a base class for inheriting classes where the domain elements can be represented by a collection of domain elements. It is represented as a bitstring where the set bits indicate that the element at position `i` are present in the domain.
- [x] Implement class `HubbardDomain` as a child class of `DiscreteDomain` where extra functionalities are added for calculating the overlap between the domain and `SpinUnresolvedONV`/`SpinResolvedONV`.

### CODE REVIEW

File: gqcp/include/Domain/DiscreteDomain.hpp [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+public:
+    /*
+     *  MARK: Constructors
+     */
+
+    /**
+     *  Create a discrete domain from a set of domain indices.
+     *
+     *  @param domain_indices             The indices that the domain contains.
+     *  @param M                            The dimension of the discrete domain, i.e. the maximum number of elements that the discrete domain can contain.
```
Comment (xdvriend) [inline gqcp/include/Domain/DiscreteDomain.hpp]:
```suggestion
     *  @param domain_indices               The indices that the domain contains.
     *  @param M                            The dimension of the discrete domain, i.e. the maximum number of elements that the discrete domain can contain.
```

File: gqcp/include/Domain/DiscreteDomain.hpp [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+    /**
+     *  @return     The number of elements the domain contains.
+     */
+    size_t numberOfElements() const { return this->domain_indices.size(); }
+
+    /**
+     *  @param rhs        The other CartesianGTO.
+     *
+     *  @return     If this discrete domain is equal to the other discrete domain.
```
Comment (xdvriend) [inline gqcp/include/Domain/DiscreteDomain.hpp]:
```suggestion
     *  @return whether this discrete domain is equal to the other discrete domain.
```

File: gqcp/include/Domain/HubbardDomain.hpp [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+     *
+     *  @param onv            The spin-resolved ONV.
+     *
+     *  @return     The number of overlapping set bits after a bit-by-bit comparison between the Hubbard domain and the spin-types of the spin-resolved ONV.
+     */
+    SpinResolved<size_t> overlapWith(const SpinResolvedONV& onv) const;
+};
+
+
+}  // namespace GQCP
```
Comment (xdvriend) [inline gqcp/include/Domain/HubbardDomain.hpp]:
```suggestion
}  // namespace GQCP

```

File: gqcp/include/Domain/SimpleDomain.hpp [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+     *  @param element        the element in the domain
+     *
+     *  @return whether the Domain contains `element'.
+     */
+    virtual bool inDomain(const ElementType& element) const = 0;
+
+    /**
+     *  @param other        the other domain
+     *
+     *  @return if this DerivedDomain is equal to the other DerivedDomain.
```
Comment (xdvriend) [inline gqcp/include/Domain/SimpleDomain.hpp]:
```suggestion
     *  @return whether this DerivedDomain is equal to the other DerivedDomain.
```

File: gqcp/include/Domain/SimpleDomain.hpp [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+     */
+    virtual bool operator==(const DerivedDomain& other) const = 0;
+
+    /**
+     *  @param other        the other domain
+     *
+     *  @return if this DerivedDomain is not equal to the other DerivedDomain.
```
Comment (xdvriend) [inline gqcp/include/Domain/SimpleDomain.hpp]:
```suggestion
     *  @return whether this DerivedDomain is not equal to the other DerivedDomain.
```

File: gqcp/src/Domain/DiscreteDomain.cpp [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+    this->domain = rhs.asBitstring();
+    this->domain_indices = rhs.domainIndices();
+
+    return (*this);
+}
+
+
+/**
+ *  @param os       The output stream which the discrete domain should be concatenated to.
+ *  @param domain      The discrete domain that should be concatenated to the output stream.
```
Comment (xdvriend) [inline gqcp/src/Domain/DiscreteDomain.cpp]:
```suggestion
 *  @param os          The output stream which the discrete domain should be concatenated to.
 *  @param domain      The discrete domain that should be concatenated to the output stream.
```

File: gqcp/tests/Domain/DiscreteDomain_test.cpp [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    // Construct a discrete domain "" of size 8 where the indices 0, 2, 4 and 6 are occupied.
+    // Note that the bitstring representation of the domain should be read from right to left.
+    const std::vector<size_t> element_indices = {0, 2, 4, 6};
+
+    GQCP::DiscreteDomain discrete_domain_from_indices(element_indices, 8);
+    BOOST_CHECK_EQUAL(discrete_domain_from_indices.asString(), std::string("10101010"));
+
+    // The same as above, but with passing an unsigned integer to the constructor.
+    // GQCP::DiscreteDomain discrete_domain_from_unsigned_int(85, 8);
+    // BOOST_CHECK_EQUAL(discrete_domain_from_unsigned_int.asString(), std::string("10101010"));
```
Comment (xdvriend) [inline gqcp/tests/Domain/DiscreteDomain_test.cpp]: This test is commented out, is that by choice or by accident?

**Review by xdvriend — APPROVED** [review]
I've added some minor changes to the documentation for sake of consistency. 

One comment is about a test being commented out.

Functionality looks good! If the comments below are addressed I think this PR is good to go!

File: gqcp/tests/Domain/DiscreteDomain_test.cpp (hunk 7 above)
Comment (dariavh) [reply gqcp/tests/Domain/DiscreteDomain_test.cpp]: Definitely by accident. The test is re-enabled in https://github.com/GQCG/GQCP/pull/1040/commits/76e69474816843229760c541144bf15177a98857.

**Pull request merged**